File: ipad-app/EigoLens/Screens/History/HistoryScreen.swift
import SwiftUI

struct HistoryScreen: View {
    var onBack: () -> Void
    @EnvironmentObject var container: AppContainer
    @StateObject private var viewModel = HistoryViewModel()
    @State private var selectedTab = 0

    var body: some View {
        VStack(spacing: 0) {
            Picker("", selection: $selectedTab) {
                Text("Recent (\(viewModel.recentHistory.count))").tag(0)
                Text("Saved (\(viewModel.bookmarks.count))").tag(1)
            }
            .pickerStyle(.segmented)
            .padding()

            if selectedTab == 0 {
                recentTab
            } else {
                bookmarksTab
            }
        }
        .navigationTitle("History")
        .navigationBarTitleDisplayMode(.large)
        .toolbar {
            ToolbarItem(placement: .navigationBarLeading) {
                Button(action: onBack) {
                    Image(systemName: "chevron.left")
                }
            }
            if selectedTab == 0 && !viewModel.recentHistory.isEmpty {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button("Clear") {
                        viewModel.clearHistory()
                    }
                    .foregroundStyle(EigoLensTheme.error)
                }
            }
        }
        .task {
            viewModel.configure(container: container)
        }
        .preferredColorScheme(.dark)
    }

    // MARK: - Recent Tab

    private var recentTab: some View {
        Group {
            if viewModel.recentHistory.isEmpty {
                emptyState(icon: "clock", message: "No recent lookups")
            } else {
                List(viewModel.recentHistory, id: \.id) { entry in
                    HStack {
                        VStack(alignment: .leading, spacing: 4) {
                            Text(entry.word)
                                .font(EigoLensTheme.titleMedium)
                            if let snippet = entry.contextSnippet {
                                Text(snippet)
                                    .font(EigoLensTheme.bodySmall)
                                    .foregroundStyle(.secondary)
                                    .lineLimit(1)
                            }
                        }

                        Spacer()

                        VStack(alignment: .trailing, spacing: 4) {
                            scopeBadgeSmall(entry.scopeLevel)
                            Text(entry.timestamp, style: .relative)
                                .font(EigoLensTheme.labelSmall)
                                .foregroundStyle(.secondary)
                        }
                    }
                }
                .listStyle(.plain)
            }
        }
    }

    // MARK: - Bookmarks Tab

    private var bookmarksTab: some View {
        Group {
            if viewModel.bookmarks.isEmpty {
                emptyState(icon: "bookmark", message: "No saved words")
            } else {
                List {
                    ForEach(viewModel.bookmarks, id: \.id) { bookmark in
                        VStack(alignment: .leading, spacing: 4) {
                            Text(bookmark.word)
                                .font(EigoLensTheme.titleMedium)
                            Text(bookmark.definition)
                                .font(EigoLensTheme.bodySmall)
                                .foregroundStyle(.secondary)
                                .lineLimit(2)
                        }
                    }
                    .onDelete { indexSet in
                        for index in indexSet {
                            let word = viewModel.bookmarks[index].word
                            viewModel.removeBookmark(word: word)
                        }
                    }
                }
                .listStyle(.plain)
            }
        }
    }

    // MARK: - Helpers

    private func emptyState(icon: String, message: String) -> some View {
        VStack(spacing: 16) {
            Spacer()
            Image(systemName: icon)
                .font(.system(size: 48))
                .foregroundStyle(.secondary)
            Text(message)
                .font(EigoLensTheme.bodyMedium)
                .foregroundStyle(.secondary)
            Spacer()
        }
        .frame(maxWidth: .infinity)
    }

    private func scopeBadgeSmall(_ scope: String) -> some View {
        Text(scope)
            .font(EigoLensTheme.labelSmall)
            .foregroundStyle(.white)
            .padding(.horizontal, 8)
            .padding(.vertical, 2)
            .background(scopeColorForString(scope), in: Capsule())
    }

    private func scopeColorForString(_ scope: String) -> Color {
        switch scope {
        case "word": return EigoLensTheme.scopeWord
        case "phrase": return EigoLensTheme.scopePhrase
        case "sentence": return EigoLensTheme.scopeSentence
        case "paragraph": return EigoLensTheme.scopeParagraph
        case "full": return EigoLensTheme.scopeFullText
        default: return .secondary
        }
    }
}

File: app/src/main/java/com/jworks/eigolens/ui/history/HistoryScreen.kt
package com.jworks.eigolens.ui.history

import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bookmark
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.DeleteSweep
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.PrimaryTabRow
import androidx.compose.material3.Tab
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import com.jworks.eigolens.R
import com.jworks.eigolens.data.local.entities.BookmarkedWordEntity
import com.jworks.eigolens.data.local.entities.LookupHistoryEntity
import com.jworks.eigolens.ui.theme.GlassBorder
import com.jworks.eigolens.ui.theme.GlassGradient
import com.jworks.eigolens.ui.theme.glassCardColors
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@Composable
fun HistoryScreen(
    onBackClick: () -> Unit,
    viewModel: HistoryViewModel = hiltViewModel()
) {
    val recentHistory by viewModel.recentHistory.collectAsState()
    val bookmarks by viewModel.bookmarks.collectAsState()
    val historyCount by viewModel.historyCount.collectAsState()
    val bookmarkCount by viewModel.bookmarkCount.collectAsState()

    var selectedTab by remember { mutableIntStateOf(0) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.background)
            .padding(top = 16.dp)
    ) {
        // Top bar
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(onClick = onBackClick) {
                Icon(
                    painter = painterResource(id = R.drawable.ic_arrow_back),
                    contentDescription = "Back",
                    tint = Color.White,
                    modifier = Modifier.size(24.dp)
                )
            }
            Text(
                text = "History",
                style = MaterialTheme.typography.headlineSmall,
                color = Color.White,
                modifier = Modifier
                    .padding(start = 8.dp)
                    .weight(1f)
            )

            if (selectedTab == 0 && recentHistory.isNotEmpty()) {
                TextButton(onClick = { viewModel.clearHistory() }) {
                    Icon(
                        Icons.Default.DeleteSweep,
                        contentDescription = null,
                        modifier = Modifier.size(18.dp)
                    )
                    Spacer(modifier = Modifier.width(4.dp))
                    Text("Clear")
                }
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        // Tab row
        @OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
        PrimaryTabRow(
            selectedTabIndex = selectedTab,
            modifier = Modifier.fillMaxWidth()
        ) {
            Tab(
                selected = selectedTab == 0,
                onClick = { selectedTab = 0 },
                text = { Text("Recent ($historyCount)") }
            )
            Tab(
                selected = selectedTab == 1,
                onClick = { selectedTab = 1 },
                text = {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Icon(
                            Icons.Default.Bookmark,
                            contentDescription = null,
                            modifier = Modifier.size(16.dp)
                        )
                        Spacer(modifier = Modifier.width(4.dp))
                        Text("Saved ($bookmarkCount)")
                    }
                }
            )
        }

        // Content
        when (selectedTab) {
            0 -> RecentTab(history = recentHistory)
            1 -> SavedTab(
                bookmarks = bookmarks,
                onRemoveBookmark = { viewModel.removeBookmark(it) }
            )
        }
    }
}

@Composable
private fun RecentTab(history: List<LookupHistoryEntity>) {
    if (history.isEmpty()) {
        EmptyState(message = "No lookups yet. Tap words in captured text to start building history.")
        return
    }

    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        items(history, key = { it.id }) { entry ->
            HistoryItem(entry)
        }
    }
}

@Composable
private fun HistoryItem(entry: LookupHistoryEntity) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .animateContentSize(),
        colors = glassCardColors(),
        border = GlassBorder,
        shape = RoundedCornerShape(12.dp)
    ) {
        Row(
            modifier = Modifier
                .background(GlassGradient)
                .fillMaxWidth()
                .padding(12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = entry.word,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
                if (entry.contextSnippet != null) {
                    Text(
                        text = entry.contextSnippet,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.padding(top = 2.dp)
                    )
                }
            }

            Spacer(modifier = Modifier.width(8.dp))

            Column(horizontalAlignment = Alignment.End) {
                ScopeBadge(scope = entry.scopeLevel)
                Text(
                    text = formatTimestamp(entry.timestamp),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(top = 2.dp)
                )
            }
        }
    }
}

@Composable
private fun SavedTab(
    bookmarks: List<BookmarkedWordEntity>,
    onRemoveBookmark: (String) -> Unit
) {
    if (bookmarks.isEmpty()) {
        EmptyState(message = "No saved words yet. Tap the bookmark icon when viewing a definition to save it.")
        return
    }

    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        items(bookmarks, key = { it.id }) { bookmark ->
            BookmarkItem(
                bookmark = bookmark,
                onRemove = { onRemoveBookmark(bookmark.word) }
            )
        }
    }
}

@Composable
private fun BookmarkItem(
    bookmark: BookmarkedWordEntity,
    onRemove: () -> Unit
) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .animateContentSize(),
        colors = glassCardColors(),
        border = GlassBorder,
        shape = RoundedCornerShape(12.dp)
    ) {
        Row(
            modifier = Modifier
                .background(GlassGradient)
                .fillMaxWidth()
                .padding(12.dp),
            verticalAlignment = Alignment.Top
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = bookmark.word,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold
                )
                if (bookmark.definition.isNotBlank()) {
                    Text(
                        text = bookmark.definition,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.padding(top = 4.dp)
                    )
                }
                if (bookmark.contextSnippet != null) {
                    Text(
                        text = bookmark.contextSnippet,
                        style = MaterialTheme.typography.bodySmall,
                        fontStyle = FontStyle.Italic,
                        color = MaterialTheme.colorScheme.outline,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.padding(top = 2.dp)
                    )
                }
                Text(
                    text = formatTimestamp(bookmark.bookmarkedAt),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(top = 4.dp)
                )
            }

            IconButton(
                onClick = onRemove,
                modifier = Modifier.size(32.dp)
            ) {
                Icon(
                    Icons.Default.Delete,
                    contentDescription = "Remove bookmark",
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.size(18.dp)
                )
            }
        }
    }
}

@Composable
private fun ScopeBadge(scope: String) {
    val color = when (scope) {
        "word" -> MaterialTheme.colorScheme.primary
        "sentence" -> MaterialTheme.colorScheme.secondary
        "phrase" -> MaterialTheme.colorScheme.tertiary
        "paragraph" -> MaterialTheme.colorScheme.error
        "full" -> MaterialTheme.colorScheme.outline
        else -> MaterialTheme.colorScheme.onSurfaceVariant
    }

    Box(
        modifier = Modifier
            .background(color.copy(alpha = 0.1f), RoundedCornerShape(6.dp))
            .padding(horizontal = 8.dp, vertical = 2.dp)
    ) {
        Text(
            text = scope,
            style = MaterialTheme.typography.labelSmall,
            color = color,
            fontWeight = FontWeight.Medium
        )
    }
}

@Composable
private fun EmptyState(message: String) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .padding(32.dp),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = message,
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = androidx.compose.ui.text.style.TextAlign.Center
        )
    }
}

private fun formatTimestamp(timestamp: Long): String {
    val now = System.currentTimeMillis()
    val diff = now - timestamp

    return when {
        diff < 60_000 -> "just now"
        diff < 3_600_000 -> "${diff / 60_000}m ago"
        diff < 86_400_000 -> "${diff / 3_600_000}h ago"
        else -> SimpleDateFormat("MMM d", Locale.getDefault()).format(Date(timestamp))
    }
}
